topo/topo.go

```diff
@@ -569,25 +570,70 @@ var (
 	new = New // a non-public new to allow overriding the New() in this package.
 )
 
+// sMap keeps the POD state of all topology nodes.
+type sMap struct {
+	m map[string]corev1.PodPhase
+}
+
+func (s *sMap) Size() int {
+	return len(s.m)
+}
+
+func (s *sMap) SetNodeState(name string, state corev1.PodPhase) {
+	if s.m == nil {
+		s.m = map[string]corev1.PodPhase{}
+	}
+	s.m[name] = state
+}
+
+func (s *sMap) TopoState() cpb.TopologyState {
+	if s == nil || len(s.m) == 0 {
+		return cpb.TopologyState_TOPOLOGY_STATE_UNKNOWN
+	}
+	cntTable := map[corev1.PodPhase]int{}
+	for _, gotState := range s.m {
+		cntTable[gotState]++
+	}
+
+	if cntTable[corev1.PodRunning] == s.Size() {
+		return cpb.TopologyState_TOPOLOGY_STATE_RUNNING
+	}
+	if cntTable[corev1.PodFailed] > 0 {
+		return cpb.TopologyState_TOPOLOGY_STATE_ERROR
+	}
+	if cntTable[corev1.PodPending] > 0 {
+		return cpb.TopologyState_TOPOLOGY_STATE_CREATING
+	}
+	return cpb.TopologyState_TOPOLOGY_STATE_UNKNOWN
+}
+
 // GetTopologyServices returns the topology information.
-func GetTopologyServices(ctx context.Context, params TopologyParams) (*tpb.Topology, error) {
+func GetTopologyServices(ctx context.Context, params TopologyParams) (*cpb.ShowTopologyResponse, error) {
 	topopb, err := Load(params.TopoName)
 	if err != nil {
-		return nil, fmt.Errorf("failed to load %s: %+v", params.TopoName, err)
+		return &cpb.ShowTopologyResponse{
```

Comment:
why are you returning both an object and an error?

Reply:
oops... done.